sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricResults.java

```diff
@@ -0,0 +1,52 @@
+package org.apache.beam.sdk.metrics;
+
+import autovalue.shaded.com.google.common.common.collect.Iterables;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.PipelineResult;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Experimental.Kind;
+
+/**
+ * Methods for interacting with the metrics of a pipeline that has been executed. Accessed via
+ * {@link PipelineResult#metrics()}.
+ */
+@Experimental(Kind.METRICS)
+public abstract class MetricResults {
+
+  /** Retrieve the current counter value. */
+  public MetricResult<Long> getCounter(MetricNameFilter name, @Nullable String step) {
```

Comment:
The same functionality can be accomplished with the more general query API, but this single-metric version is a nice lightweight syntax for the single-metric case. So looking at it from the usage perspective:

```java
Pipeline p = Pipeline.create(..);
MetricName fooSize = MetricName.create(MyPipeline.class, "fooSize");
p.apply(..)
  .apply(MapElements.via(foo => {
    fooSize.add(foo.size());
    // ..
  })
  .apply(..);

PipelineResult result = p.run();
while (true) {
  Thread.sleep(1000);
  LOG.info("Mean foo size: {}", result.metrics().getDistribution(fooSize).committed().mean());
}
```

- I like the ability to not specify a step name, otherwise using metrics also requires naming step PTransform application.
- Providing a `MetricName` instead of a `MetricNameFilter` seems reasonable, since we really only want equality matching on a `MetricName` and not any other semantics `MetricNameFilter` may provide.